cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/TaskMetricsWriter.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright © 2017 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.internal.app.runtime.batch;
+
+import co.cask.cdap.api.metrics.MetricsContext;
+import co.cask.cdap.app.metrics.MapReduceMetrics;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.internal.app.runtime.AbstractContext;
+import org.apache.hadoop.mapreduce.TaskCounter;
+import org.apache.hadoop.mapreduce.TaskInputOutputContext;
+
+import java.util.Map;
+
+/**
+ * Base class for gathering statistics from a running map/reduce task through its counters and for writing the data to
+ * the metrics system.
+ */
+public abstract class TaskMetricsWriter {
+  public static final String PROPERTY_NAME = "c.mr.task.metrics.emit.interval.in.secs";
+  private final TaskInputOutputContext taskContext;
+  private final MetricsContext metricsContext;
+
+  TaskMetricsWriter(AbstractContext context, TaskInputOutputContext taskContext) {
+    this.taskContext = taskContext;
+    this.metricsContext = context.getProgramMetrics();
```

Comment:
If you only need the `MetricsContext`, have it pass from the constructor directly. This make the class more generic and easier to use/test/mock.